### PR TITLE
Introduce Bpf Program Snapshot (bps)

### PR DESCRIPTION
**Pull request by iamkafai**
This patch set introduces a new binary command/tool to list all loaded BPF programs of a system.  The first two patches are prep work.  The last patch introduces the 'bps' binary and its commit message has some sample output.

### CODE REVIEW

**Comment by yonghong-song** [comment]
[buildbot, ok to test]

**Comment by yonghong-song** [comment]
@drzaeus77 could you help check buildbot? Seems not working now?

File: introspection/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) Facebook, Inc.
```
Comment (yonghong-song) [inline introspection/CMakeLists.txt]: Year "2017"?

File: introspection/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+# Licensed under the Apache License, Version 2.0 (the "License")
+
+include_directories(${CMAKE_SOURCE_DIR}/src/cc)
+include_directories(${CMAKE_SOURCE_DIR}/src/cc/api)
+
+option(INSTALL_INTROSPECTION "Install BPF introspection tools. Those binaries are statically linked and can take plenty of disk space" ON)
```
Comment (yonghong-song) [inline introspection/CMakeLists.txt]: It is static linking. However, all llvm related functions are not called and hence not included in binary. So currently, the size is actually fair small.

-bash-4.3$ ls -l /usr/share/bcc/introspection/bps 
-rwxr-xr-x. 1 root root 37872 Oct 21 03:44 /usr/share/bcc/introspection/bps
-bash-4.3$ 

So I suggest to remove the sentence "Those binaries are statically linked and can take plenty of space."

**Comment by yonghong-song** [comment]
[buildbot, ok to test]

**Comment by palmtenor** [comment]
Do you think we should add `deprecated` attribute to the old `bpf_map_create` and `bpf_prog_load` without name? This would encourage people to move to the new helper and provide better information for introspection

**Comment by iamkafai** [comment]
@palmtenor , how about I remove bpf_prog_load_name and bpf_create_map_name? and just add a 'const char *name' argument to the existing bpf_prog_name() and bpf_create_map()?  That we can force people to change.

**Comment by iamkafai** [comment]
Last set of changes:
1. Fix rpmbuild error
2. Fix tests/python/test_trace4.py test case to accommodate the extra retry-call to syscall(__NR_bpf, BPF_PROG_LOAD) when the kernel does not support prog_name
3. After discussing with @yonghong-song and @palmtenor , my understand is that it is acceptable to change the libbpf's ABI in this stage.  The preference is to keep a clean API by having one bpf_prog_load() instead of adding another _name() variants.  This patch made this change by adding an extra 'const char *name' argument to bpf_prog_load() and bpf_create_map().
4. Due to (3), the lua side has to change also.  I have tested ./bcc-probe ../../examples/lua/task_switch.lua

**Comment by yonghong-song** [comment]
[buildbot, ok to test]

**Pull request merged**